File: apps/windows/Lumina/Features/Reader/Listen/ListenScript.cs
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Lumina.Features.Reader.Listen;

public enum ListenMode
{
    Summary,
    Detailed,
    Original,
}

public abstract record ListenHighlightAnchor
{
    public sealed record SegmentTitle : ListenHighlightAnchor;
    public sealed record SummarySentence(int Index) : ListenHighlightAnchor;
    public sealed record SectionBullets : ListenHighlightAnchor;
    public sealed record Bullet(int Index) : ListenHighlightAnchor;
    public sealed record OriginalUtf16(int Start, int Length) : ListenHighlightAnchor;
}

public sealed record ListenUtterance(string Text, ListenHighlightAnchor Anchor);

/// Policy for listen follow-along UI side effects.
public static class ListenFollowHighlightPolicy
{
    /// Must stay false: auto-scrolling spoken lines fights continuous play advance.
    public const bool ScrollsUtteranceIntoView = false;
}

public sealed record ListenScript(
    ListenMode Mode,
    string Language,
    IReadOnlyList<ListenUtterance> Utterances,
    bool Ready,
    string? SkipReason)
{
    public const string SectionBullets = "主要内容";
    public const string SectionNotes = "需要注意";
    public const int MaxUtteranceChars = 800;

    public IReadOnlyList<string> Texts => Utterances.Select(u => u.Text).ToList();

    public static ListenScript NotReady(ListenMode mode, string reason, string language = "zh") =>
        new(mode, language, [], false, reason);
}

/// Whether / how to announce chapter before segment label (PRD §5.3.1).
public abstract record ListenChapterSpeakContext
{
    public sealed record SessionStart : ListenChapterSpeakContext;
    public sealed record Continuing(string? PreviousSpokenChapter) : ListenChapterSpeakContext;
}

public static class ListenChapterAnnouncePolicy
{
    public static string? NormalizedChapter(string? raw)
    {
        var cleaned = SegmentCatalogPolicy.StripSectionMark(raw ?? "");
        return cleaned.Length == 0 ? null : cleaned;
    }

    public static bool ShouldAnnounce(string? chapter, ListenChapterSpeakContext context)
    {
        var current = NormalizedChapter(chapter);
        if (current is null) return false;
        return context switch
        {
            ListenChapterSpeakContext.SessionStart => true,
            ListenChapterSpeakContext.Continuing cont =>
                NormalizedChapter(cont.PreviousSpokenChapter) != current,
            _ => false,
        };
    }

    public static List<string> TitlePrefix(
        string? segmentLabel,
        string? chapter,
        ListenChapterSpeakContext context)
    {
        var lines = new List<string>();
        var chapterName = NormalizedChapter(chapter);
        var announce = ShouldAnnounce(chapter, context);
        if (announce && chapterName is not null)
            lines.Add(chapterName);
        var title = ListenScriptBuilder.SegmentTitle(segmentLabel);
        if (title is not null)
        {
            if (title != chapterName)
                lines.Add(title);
            else if (!announce)
                lines.Add(title);
        }
        return lines;
    }
}

public static class ListenScriptBuilder
{
    private static readonly Regex CjkRe = new(@"[\u4e00-\u9fff]", RegexOptions.Compiled);

    public static ListenMode ParseMode(string? raw)
    {
        return (raw ?? "").Trim().ToLowerInvariant() switch
        {
            "detailed" => ListenMode.Detailed,
            "original" => ListenMode.Original,
            _ => ListenMode.Summary,
        };
    }

    public static string Label(ListenMode mode) => mode switch
    {
        ListenMode.Detailed => "听完整摘要",
        ListenMode.Original => "听原文",
        _ => "听简要摘要",
    };

    public static string ShortLabel(ListenMode mode) => mode switch
    {
        ListenMode.Detailed => "完整摘要",
        ListenMode.Original => "原文",
        _ => "简要摘要",
    };

    public static bool IsSummaryLayer(ListenMode mode) =>
        mode is ListenMode.Summary or ListenMode.Detailed;

    public static string DetectLanguage(string text)
    {
        var trimmed = text.Trim();
        if (trimmed.Length == 0) return "zh";
        var cjk = CjkRe.Matches(trimmed).Count;
        var letters = trimmed.Count(ch => (ch is >= 'A' and <= 'Z') || (ch is >= 'a' and <= 'z'));
        return cjk >= Math.Max(1, letters) ? "zh" : "en";
    }

    public static ListenScript Build(
        ListenMode mode,
        string? summaryJson,
        string? rawText,
        string? languageHint = null,
        string? segmentLabel = null,
        string? chapter = null,
        ListenChapterSpeakContext? chapterContext = null)
    {
        chapterContext ??= new ListenChapterSpeakContext.SessionStart();
        var prefix = ListenChapterAnnouncePolicy.TitlePrefix(segmentLabel, chapter, chapterContext);
        if (mode == ListenMode.Original)
        {
            var source = rawText ?? "";
            var text = source.Trim();
            if (text.Length == 0)
                return ListenScript.NotReady(ListenMode.Original, "empty_text", languageHint ?? "zh");
            var utterances = new List<ListenUtterance>();
            foreach (var line in prefix)
                utterances.AddRange(AnchoredChunks(line, new ListenHighlightAnchor.SegmentTitle()));
            utterances.AddRange(OriginalUtterances(source));
            var language = languageHint ?? DetectLanguage(text);
            if (utterances.Count == 0)
                return ListenScript.NotReady(ListenMode.Original, "empty_text", language);
            return new ListenScript(ListenMode.Original, language, utterances, true, null);
        }

        if (!TryParseSummary(summaryJson, out var sentences, out var bullets, out var notes))
            return ListenScript.NotReady(mode, "summary_not_ready", languageHint ?? "zh");

        var utterancesSummary = new List<ListenUtterance>();
        foreach (var line in prefix)
            utterancesSummary.AddRange(AnchoredChunks(line, new ListenHighlightAnchor.SegmentTitle()));
        for (var i = 0; i < sentences.Count; i++)
        {
            var sentence = sentences[i].Trim();
            if (sentence.Length == 0) continue;
            utterancesSummary.AddRange(AnchoredChunks(sentence, new ListenHighlightAnchor.SummarySentence(i)));
        }
        if (mode == ListenMode.Detailed)
        {
            var kept = new List<(int SourceIndex, (string Label, string Body) Bullet)>();
            for (var i = 0; i < bullets.Count; i++)
            {
                if (bullets[i].Body.Trim().Length == 0) continue;
                kept.Add((i, bullets[i]));
            }
            if (kept.Count > 0)
            {
                utterancesSummary.AddRange(AnchoredChunks(
                    ListenScript.SectionBullets, new ListenHighlightAnchor.SectionBullets()));
                for (var display = 0; display < kept.Count; display++)
                {
                    var (sourceIndex, bullet) = kept[display];
                    var line = FormatBullet(display + 1, bullet.Label, bullet.Body);
                    utterancesSummary.AddRange(AnchoredChunks(
                        line, new ListenHighlightAnchor.Bullet(sourceIndex)));
                }
            }
        }

        var sampleParts = new List<string>();
        sampleParts.AddRange(prefix);
        sampleParts.AddRange(sentences);
        sampleParts.AddRange(bullets.Select(b => b.Body));
        sampleParts.AddRange(notes);
        var sample = string.Join(" ", sampleParts);
        var lang = languageHint ?? DetectLanguage(sample);
        if (utterancesSummary.Count == 0)
            return ListenScript.NotReady(mode, "summary_not_ready", lang);
        return new ListenScript(mode, lang, utterancesSummary, true, null);
    }

    /// Condensed segment title only; blank means skip (no 段 N fallback).
    public static string? SegmentTitle(string? segmentLabel)
    {
        var cleaned = (segmentLabel ?? "").Trim();
        return cleaned.Length == 0 ? null : cleaned;
    }

    public static string FormatBullet(int index, string? label, string body)
    {
        if (!string.IsNullOrEmpty(label))
            return $"{index}. {label}。{body}";
        return $"{index}. {body}";
    }

    public static List<string> SplitSentences(string text)
    {
        var raw = text.Trim();
        if (raw.Length == 0) return [];
        var result = new List<string>();
        var current = new StringBuilder();
        var chars = raw.ToCharArray();
        var i = 0;
        var cnStops = new HashSet<char> { '。', '！', '？', '；' };
        while (i < chars.Length)
        {
            var ch = chars[i];
            if (ch == '\n')
            {
                var piece = current.ToString().Trim();
                if (piece.Length > 0) result.Add(piece);
                current.Clear();
                i += 1;
                continue;
            }
            current.Append(ch);
            var isCn = cnStops.Contains(ch);
            var isBang = ch is '!' or '?';
            var isPeriod = ch == '.' && (i + 1 >= chars.Length || !char.IsDigit(chars[i + 1]));
            if (isCn || isBang || isPeriod)
            {
                var j = i + 1;
                while (j < chars.Length && char.IsWhiteSpace(chars[j])) j += 1;
                var piece = current.ToString().Trim();
                if (piece.Length > 0) result.Add(piece);
                current.Clear();
                i = j;
                continue;
            }
            i += 1;
        }
        var tail = current.ToString().Trim();
        if (tail.Length > 0) result.Add(tail);
        return result.Count == 0 ? [raw] : result;
    }

    public static List<string> ChunkLong(string text, int maxChars = ListenScript.MaxUtteranceChars)
    {
        var cleaned = Regex.Replace(text.Trim(), @"\s+", " ");
        if (cleaned.Length == 0) return [];
        if (cleaned.Length <= maxChars) return [cleaned];
        var chunks = new List<string>();
        var remaining = cleaned;
        while (remaining.Length > 0)
        {
            if (remaining.Length <= maxChars)
            {
                chunks.Add(remaining);
                break;
            }
            var window = remaining[..maxChars];
            var cut = LastBreak(window) ?? maxChars;
            cut = Math.Min(cut, remaining.Length);
            var piece = remaining[..cut].Trim();
            if (piece.Length > 0) chunks.Add(piece);
            remaining = remaining[cut..].Trim();
        }
        return chunks;
    }

    private static List<ListenUtterance> AnchoredChunks(string text, ListenHighlightAnchor anchor)
    {
        return ChunkLong(text).Select(chunk => new ListenUtterance(chunk, anchor)).ToList();
    }

    private static List<ListenUtterance> OriginalUtterances(string source)
    {
        var outList = new List<ListenUtterance>();
        var searchFrom = 0;
        foreach (var piece in SplitSentences(source))
        {
            var located = LocateRange(piece, source, ref searchFrom);
            var chunks = ChunkLong(piece);
            if (located is null)
            {
                outList.AddRange(AnchoredChunks(piece, new ListenHighlightAnchor.OriginalUtf16(0, 0)));
                continue;
            }
            var (start, length) = located.Value;
            foreach (var chunk in chunks)
                outList.Add(new ListenUtterance(chunk, new ListenHighlightAnchor.OriginalUtf16(start, length)));
        }
        return outList;
    }

    private static (int Start, int Length)? LocateRange(string needle, string haystack, ref int searchFrom)
    {
        if (searchFrom < 0) searchFrom = 0;
        if (searchFrom > haystack.Length) searchFrom = 0;
        var idx = haystack.IndexOf(needle, searchFrom, StringComparison.Ordinal);
        if (idx < 0)
            idx = haystack.IndexOf(needle, StringComparison.Ordinal);
        if (idx < 0) return null;
        searchFrom = idx + needle.Length;
        return (idx, needle.Length);
    }

    private static int? LastBreak(string window)
    {
        string[] marks = ["。", "！", "？", "；", "，", ". ", " "];
        int? best = null;
        foreach (var mark in marks)
        {
            var idx = window.LastIndexOf(mark, StringComparison.Ordinal);
            if (idx < 0) continue;
            var end = idx + mark.Length;
            if (end >= window.Length / 3)
                best = Math.Max(best ?? 0, end);
        }
        return best;
    }

    private static bool TryParseSummary(
        string? summaryJson,
        out List<string> sentences,
        out List<(string Label, string Body)> bullets,
        out List<string> notes)
    {
        sentences = [];
        bullets = [];
        notes = [];
        if (string.IsNullOrWhiteSpace(summaryJson)) return false;
        try
        {
            using var doc = JsonDocument.Parse(summaryJson);
            if (doc.RootElement.ValueKind != JsonValueKind.Object) return false;
            var root = doc.RootElement;
            sentences = StringList(root, "sentences");
            bullets = BulletList(root, "bullets");
            notes = StringList(root, "notes");
            return sentences.Count > 0 || bullets.Count > 0 || notes.Count > 0;
        }
        catch (JsonException)
        {
            return false;
        }
    }

    private static List<string> StringList(JsonElement root, string name)
    {
        if (!root.TryGetProperty(name, out var el) || el.ValueKind != JsonValueKind.Array)
            return [];
        var outList = new List<string>();
        foreach (var item in el.EnumerateArray())
        {
            if (item.ValueKind == JsonValueKind.String)
            {
                var value = SummaryJsonParser.CollapseProseWhitespace(item.GetString());
                if (value.Length > 0) outList.Add(value);
            }
        }
        return outList;
    }

    private static List<(string Label, string Body)> BulletList(JsonElement root, string name)
    {
        if (!root.TryGetProperty(name, out var el) || el.ValueKind != JsonValueKind.Array)
            return [];
        var outList = new List<(string, string)>();
        foreach (var item in el.EnumerateArray())
        {
            if (item.ValueKind == JsonValueKind.Object)
            {
                var label = Prop(item, "label");
                var body = Prop(item, "body");
                if (body.Length == 0) body = Prop(item, "content");
                if (body.Length == 0) body = Prop(item, "text");
                if (body.Length == 0 && label.Length > 0)
                {
                    body = label;
                    label = "";
                }
                if (body.Length > 0) outList.Add((label, body));
            }
            else if (item.ValueKind == JsonValueKind.String)
            {
                var value = SummaryJsonParser.CollapseProseWhitespace(item.GetString());
                if (value.Length > 0) outList.Add(("", value));
            }
        }
        return outList;
    }

    private static string Prop(JsonElement obj, string name)
    {
        if (!obj.TryGetProperty(name, out var el)) return "";
        return el.ValueKind == JsonValueKind.String
            ? SummaryJsonParser.CollapseProseWhitespace(el.GetString())
            : "";
    }
}

/// Speaker click follows the panel on screen. Windows uses a global 原文 toggle;
/// macOS also has per-segment 切换原文 / 切换摘要, encoded as the expanded flags.
public static class ListenChromePolicy
{
    public static bool IsShowingOriginal(bool originalLayer, bool sourceExpanded, bool summaryExpanded)
    {
        if (originalLayer) return !summaryExpanded;
        return sourceExpanded;
    }

    public static ListenMode PrimaryMode(bool showingOriginal) =>
        showingOriginal ? ListenMode.Original : ListenMode.Summary;

    public static bool ShowsSummaryChevron(bool showingOriginal) => !showingOriginal;
}

File: apps/macos/Lumina/Features/Reader/Listen/ListenScript.swift
import Foundation

/// Speaker click follows the panel on screen, not only the 摘要|原文 picker.
/// Per-segment 「切换原文 / 切换摘要」 can diverge from the picker; listen must match.
enum ListenChromePolicy {
    static func isShowingOriginal(
        contentMode: ReaderContentMode,
        sourceExpanded: Bool,
        summaryExpanded: Bool
    ) -> Bool {
        switch contentMode {
        case .summary: sourceExpanded
        case .original: !summaryExpanded
        }
    }

    static func primaryMode(showingOriginal: Bool) -> ListenMode {
        showingOriginal ? .original : .summary
    }

    static func showsSummaryChevron(showingOriginal: Bool) -> Bool {
        !showingOriginal
    }
}

enum ListenMode: String, CaseIterable, Equatable {
    case summary
    case detailed
    case original

    var label: String {
        switch self {
        case .summary: "听简要摘要"
        case .detailed: "听完整摘要"
        case .original: "听原文"
        }
    }

    var shortLabel: String {
        switch self {
        case .summary: "简要摘要"
        case .detailed: "完整摘要"
        case .original: "原文"
        }
    }

    var isSummaryLayer: Bool {
        self == .summary || self == .detailed
    }
}

/// Where the UI should paint the light follow-along highlight for one TTS utterance.
enum ListenHighlightAnchor: Equatable {
    case segmentTitle
    case summarySentence(Int)
    case sectionBullets
    case bullet(Int)
    /// UTF-16 range into the segment `raw_text` shown in the reader.
    case originalUTF16(location: Int, length: Int)

    var originalNSRange: NSRange? {
        guard case let .originalUTF16(location, length) = self, length > 0 else { return nil }
        return NSRange(location: location, length: length)
    }
}

/// Whether / how to announce chapter before segment label (PRD §5.3.1).
enum ListenChapterSpeakContext: Equatable {
    /// First segment of this listen session (or fresh start).
    case sessionStart
    /// After at least one segment was spoken; compare to last spoken chapter.
    case continuing(previousSpokenChapter: String?)
}

enum ListenChapterAnnouncePolicy {
    static func normalizedChapter(_ raw: String?) -> String? {
        let cleaned = SegmentOutlinePolicy.stripSectionMark(raw ?? "")
        return cleaned.isEmpty ? nil : cleaned
    }

    static func shouldAnnounce(chapter: String?, context: ListenChapterSpeakContext) -> Bool {
        guard let current = normalizedChapter(chapter) else { return false }
        switch context {
        case .sessionStart:
            return true
        case .continuing(let previous):
            return normalizedChapter(previous) != current
        }
    }

    /// Ordered title lines before body: optional chapter, then label (skip duplicate).
    static func titlePrefix(
        segmentLabel: String?,
        chapter: String?,
        context: ListenChapterSpeakContext
    ) -> [String] {
        var lines: [String] = []
        let chapterName = normalizedChapter(chapter)
        let announce = shouldAnnounce(chapter: chapter, context: context)
        if announce, let chapterName {
            lines.append(chapterName)
        }
        if let title = ListenScript.segmentTitle(segmentLabel) {
            if title != chapterName {
                lines.append(title)
            } else if !announce {
                lines.append(title)
            }
        }
        return lines
    }
}

struct ListenUtterance: Equatable {
    var text: String
    var anchor: ListenHighlightAnchor
}

struct ListenScript: Equatable {
    var mode: ListenMode
    var language: String
    var utterances: [ListenUtterance]
    var ready: Bool
    var skipReason: String?

    static let sectionBullets = "主要内容"
    static let sectionNotes = "需要注意"
    static let maxUtteranceChars = 800

    var texts: [String] { utterances.map(\.text) }

    static func notReady(_ mode: ListenMode, reason: String, language: String = "zh") -> ListenScript {
        ListenScript(mode: mode, language: language, utterances: [], ready: false, skipReason: reason)
    }

    static func detectLanguage(_ text: String) -> String {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return "zh" }
        var cjk = 0
        var letters = 0
        for scalar in trimmed.unicodeScalars {
            if (0x4E00...0x9FFF).contains(scalar.value) {
                cjk += 1
            } else if (0x41...0x5A).contains(scalar.value) || (0x61...0x7A).contains(scalar.value) {
                letters += 1
            }
        }
        return cjk >= max(1, letters) ? "zh" : "en"
    }

    static func build(
        mode: ListenMode,
        summary: ParsedSummary?,
        rawText: String?,
        languageHint: String? = nil,
        segmentLabel: String? = nil,
        chapter: String? = nil,
        chapterContext: ListenChapterSpeakContext = .sessionStart
    ) -> ListenScript {
        let prefix = ListenChapterAnnouncePolicy.titlePrefix(
            segmentLabel: segmentLabel,
            chapter: chapter,
            context: chapterContext
        )
        switch mode {
        case .original:
            let source = rawText ?? ""
            let trimmed = source.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmed.isEmpty else {
                return notReady(.original, reason: "empty_text", language: languageHint ?? "zh")
            }
            var utterances: [ListenUtterance] = []
            for line in prefix {
                utterances.append(contentsOf: anchoredChunks(line, anchor: .segmentTitle))
            }
            utterances.append(contentsOf: originalUtterances(in: source))
            let language = languageHint ?? detectLanguage(trimmed)
            if utterances.isEmpty {
                return notReady(.original, reason: "empty_text", language: language)
            }
            return ListenScript(mode: .original, language: language, utterances: utterances, ready: true, skipReason: nil)
        case .summary, .detailed:
            guard let summary, summary.hasContent else {
                return notReady(mode, reason: "summary_not_ready", language: languageHint ?? "zh")
            }
            var utterances: [ListenUtterance] = []
            for line in prefix {
                utterances.append(contentsOf: anchoredChunks(line, anchor: .segmentTitle))
            }
            for (index, sentence) in summary.sentences.enumerated() {
                let cleaned = sentence.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !cleaned.isEmpty else { continue }
                utterances.append(contentsOf: anchoredChunks(cleaned, anchor: .summarySentence(index)))
            }
            if mode == .detailed {
                let bullets = summary.bullets.enumerated().compactMap { index, bullet -> (Int, ParsedBullet)? in
                    let body = bullet.body.trimmingCharacters(in: .whitespacesAndNewlines)
                    guard !body.isEmpty else { return nil }
                    return (index, bullet)
                }
                if !bullets.isEmpty {
                    utterances.append(contentsOf: anchoredChunks(sectionBullets, anchor: .sectionBullets))
                    for (displayIndex, (sourceIndex, bullet)) in bullets.enumerated() {
                        let line = formatBullet(index: displayIndex + 1, label: bullet.label, body: bullet.body)
                        utterances.append(contentsOf: anchoredChunks(line, anchor: .bullet(sourceIndex)))
                    }
                }
            }
            let sample = (prefix + summary.sentences + summary.bullets.map(\.body) + summary.notes)
                .joined(separator: " ")
            let language = languageHint ?? detectLanguage(sample)
            if utterances.isEmpty {
                return notReady(mode, reason: "summary_not_ready", language: language)
            }
            return ListenScript(mode: mode, language: language, utterances: utterances, ready: true, skipReason: nil)
        }
    }

    /// Condensed segment title only; blank means skip (no 段 N fallback).
    static func segmentTitle(_ segmentLabel: String?) -> String? {
        let cleaned = segmentLabel?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return cleaned.isEmpty ? nil : cleaned
    }

    static func formatBullet(index: Int, label: String?, body: String) -> String {
        if let label, !label.isEmpty {
            return "\(index). \(label)。\(body)"
        }
        return "\(index). \(body)"
    }

    static func splitSentences(_ text: String) -> [String] {
        splitSentencePieces(text).map(\.text)
    }

    private static func originalUtterances(in source: String) -> [ListenUtterance] {
        var out: [ListenUtterance] = []
        var searchFrom = source.startIndex
        for piece in splitSentencePieces(source) {
            guard let located = locateUTF16Range(of: piece.text, in: source, from: &searchFrom) else {
                // Speak without a paint range if the substring cannot be relocated.
                out.append(contentsOf: anchoredChunks(
                    piece.text,
                    anchor: .originalUTF16(location: 0, length: 0)
                ))
                continue
            }
            let chunks = chunkLong(piece.text)
            if chunks.count <= 1 {
                out.append(ListenUtterance(text: chunks.first ?? piece.text, anchor: .originalUTF16(
                    location: located.location,
                    length: located.length
                )))
                continue
            }
            // Long sentences: keep the whole sentence range highlighted across chunks.
            for chunk in chunks {
                out.append(ListenUtterance(
                    text: chunk,
                    anchor: .originalUTF16(location: located.location, length: located.length)
                ))
            }
        }
        return out
    }

    private static func anchoredChunks(_ text: String, anchor: ListenHighlightAnchor) -> [ListenUtterance] {
        chunkLong(text).map { ListenUtterance(text: $0, anchor: anchor) }
    }

    private struct SentencePiece {
        var text: String
    }

    private static func splitSentencePieces(_ text: String) -> [SentencePiece] {
        let raw = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !raw.isEmpty else { return [] }
        var result: [SentencePiece] = []
        var current = ""
        let chars = Array(raw)
        var i = 0
        let cnStops = Set("。！？；")
        while i < chars.count {
            let ch = chars[i]
            if ch == "\n" {
                let piece = current.trimmingCharacters(in: .whitespacesAndNewlines)
                if !piece.isEmpty { result.append(SentencePiece(text: piece)) }
                current = ""
                i += 1
                continue
            }
            current.append(ch)
            let isCN = cnStops.contains(ch)
            let isBang = ch == "!" || ch == "?"
            let isPeriod = ch == "." && (i + 1 >= chars.count || !chars[i + 1].isNumber)
            if isCN || isBang || isPeriod {
                var j = i + 1
                while j < chars.count && chars[j].isWhitespace { j += 1 }
                let piece = current.trimmingCharacters(in: .whitespacesAndNewlines)
                if !piece.isEmpty { result.append(SentencePiece(text: piece)) }
                current = ""
                i = j
                continue
            }
            i += 1
        }
        let tail = current.trimmingCharacters(in: .whitespacesAndNewlines)
        if !tail.isEmpty { result.append(SentencePiece(text: tail)) }
        return result.isEmpty ? [SentencePiece(text: raw)] : result
    }

    private static func locateUTF16Range(
        of needle: String,
        in haystack: String,
        from searchFrom: inout String.Index
    ) -> NSRange? {
        if searchFrom > haystack.endIndex { searchFrom = haystack.startIndex }
        if searchFrom < haystack.startIndex { searchFrom = haystack.startIndex }
        if let range = haystack.range(of: needle, range: searchFrom..<haystack.endIndex) {
            searchFrom = range.upperBound
            return NSRange(range, in: haystack)
        }
        if let range = haystack.range(of: needle) {
            searchFrom = range.upperBound
            return NSRange(range, in: haystack)
        }
        return nil
    }

    static func chunkLong(_ text: String, maxChars: Int = maxUtteranceChars) -> [String] {
        let cleaned = text.trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
        guard !cleaned.isEmpty else { return [] }
        if cleaned.count <= maxChars { return [cleaned] }
        var chunks: [String] = []
        var remaining = cleaned
        while !remaining.isEmpty {
            if remaining.count <= maxChars {
                chunks.append(remaining)
                break
            }
            let end = remaining.index(remaining.startIndex, offsetBy: maxChars)
            let window = String(remaining[..<end])
            let cut = lastBreak(in: window) ?? maxChars
            let cutIndex = remaining.index(remaining.startIndex, offsetBy: min(cut, remaining.count))
            let piece = remaining[..<cutIndex].trimmingCharacters(in: .whitespacesAndNewlines)
            if !piece.isEmpty { chunks.append(String(piece)) }
            remaining = remaining[cutIndex...].trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return chunks
    }

    private static func lastBreak(in window: String) -> Int? {
        let marks = ["。", "！", "？", "；", "，", ". ", " "]
        var best: Int?
        for mark in marks {
            if let range = window.range(fromEnd: mark) {
                let idx = window.distance(from: window.startIndex, to: range.upperBound)
                if idx >= window.count / 3 {
                    best = max(best ?? 0, idx)
                }
            }
        }
        return best
    }
}

private extension String {
    func range(fromEnd search: String) -> Range<String.Index>? {
        range(of: search, options: .backwards)
    }
}
